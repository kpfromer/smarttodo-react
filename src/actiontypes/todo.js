export const REQUEST_TODOS = 'todo/REQUEST_TODOS';
export const REQUEST_FAILURE_TODOS = 'todo/REQUEST_FAILURE_TODOS';
export const RECIEVE_TODOS = 'todo/RECIEVE_TODOS';

export const ADD_TODO = 'todo/ADD_TODO';
export const ADD_TODO_FAILURE = 'todo/ADD_TODO_FAILURE';
export const ADD_TODO_SUCCESS = 'todo/ADD_TODO_SUCCESS';

export const UPDATE_TODO = 'todo/UPDATE_TODO';
export const UPDATE_TODO_SUCCESS = 'todo/UPDATE_TODO_SUCCESS';
export const UPDATE_TODO_FAILURE = 'todo/UPDATE_TODO_FAILURE';

export const REMOVE_TODO = 'todo/REMOVE_TODO';
export const REMOVE_TODO_SUCCESS = 'todo/REMOVE_TODO_SUCCESS';
export const REMOVE_TODO_FAILURE = 'todo/REMOVE_TODO_FAILURE';