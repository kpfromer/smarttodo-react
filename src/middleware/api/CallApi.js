import { API_ROOT } from "./api";
import axios from 'axios';

export default (endpoint, method, mapResponse = value => value, data = undefined, token = undefined) => {

  const fullUrl = endpoint.includes(API_ROOT) ? endpoint : API_ROOT + endpoint;

  return axios({
    url: fullUrl,
    method,
    headers: {
      'Content-Type': 'application/json',
      'Authorization': token ? `Bearer ${token}` : undefined
    },
    data
  })
    .then(response => mapResponse(response.data))
    .catch(error => Promise.reject(error));
};
