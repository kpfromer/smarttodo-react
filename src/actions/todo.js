import * as TodoActionTypes from '../actiontypes/todo';
import uuid from 'uuid/v4';
import { CALL_API } from "../middleware/api/api";

export const fetchTodos = () => ({
  [CALL_API]: {
    types: [
      TodoActionTypes.REQUEST_TODOS,
      TodoActionTypes.RECIEVE_TODOS,
      TodoActionTypes.REQUEST_FAILURE_TODOS
    ],
    method: 'GET',
    endpoint: '/todo',
    minimumDelay: 1000
  }
});

export const addTodo = (description, completed) => {
  const todo = { description, completed };
  return {
    tempId: uuid(),
    todo,
    [CALL_API]: {
      types: [
        TodoActionTypes.ADD_TODO,
        TodoActionTypes.ADD_TODO_SUCCESS,
        TodoActionTypes.ADD_TODO_FAILURE
      ],
      method: 'POST',
      endpoint: '/todo',
      data: todo
    }
  }
};

export const updateTodo = (prevTodo, newTodo, id) => {

  const { id: unUsedId, ...newTodoWithoutId } = newTodo;

  return {
    id,
    todo: newTodo,
    revert: {
      todo: prevTodo
    },
    [CALL_API]: {
      types: [
        TodoActionTypes.UPDATE_TODO,
        TodoActionTypes.UPDATE_TODO_SUCCESS,
        TodoActionTypes.UPDATE_TODO_FAILURE
      ],
      method: 'PUT',
      endpoint: `/todo/${id}`,
      data: newTodoWithoutId
    }
  }
};

export const removeTodo = (todo, position, id) => ({
  id,
  revert: {
    todo,
    position
  },
  [CALL_API]: {
    types: [
      TodoActionTypes.REMOVE_TODO,
      TodoActionTypes.REMOVE_TODO_SUCCESS,
      TodoActionTypes.REMOVE_TODO_FAILURE
    ],
    method: 'DELETE',
    endpoint: `/todo/${id}`
  }
});
