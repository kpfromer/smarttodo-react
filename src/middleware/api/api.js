import CallApi from './CallApi';

export const API_ROOT = 'http://localhost:3001/v1';
export const CALL_API = 'call/CALL_API';

const CREATE_TYPES = [
  'POST',
  'PUT'
];
const VALID_TYPES = [
  ...CREATE_TYPES,
  'GET',
  'DELETE'
];

export default ({ getAuthFromState } = { getAuthFromState: val => val }) => store => next => action => {
  const callAPI = action[CALL_API];

  if (typeof callAPI === 'undefined') {
    return next(action);
  }

  let { endpoint } = callAPI;
  const { types, mapResponse, method, data, authenticate = true } = callAPI;

  if (typeof endpoint === 'function') {
    endpoint = endpoint(store.getState())
  }

  if (typeof endpoint !== 'string') {
    throw new Error('Specify a string endpoint URL.')
  }
  if (!Array.isArray(types) || types.length !== 3) {
    throw new Error('Expected an array of three action types.')
  }
  if (!types.every(type => typeof type === 'string')) {
    throw new Error('Expected action types to be strings.')
  }

  if (!VALID_TYPES.includes(method)) {
    throw new Error(`Invalid call api method: ${method}`);
  }

  if (CREATE_TYPES.includes(method) && !data) {
    throw new Error(`Method: ${method} requires data. Data is currently: ${data}`);
  }

  const actionWith = data => {
    const finalAction = { ...action, ...data };
    delete finalAction[CALL_API];
    return finalAction;
  };

  let token;

  if (authenticate) {
    token = getAuthFromState(store.getState());
  }

  const [requestType, successType, failureType] = types;
  next(actionWith({ type: requestType }));

  return CallApi(endpoint, method, mapResponse, data, token).then(
    response => next(actionWith({
      response,
      type: successType
    })),
    error => next(actionWith({
      type: failureType,
      error: error.message || 'Something bad happened'
    }))
  )
};
