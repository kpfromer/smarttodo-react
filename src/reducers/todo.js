import * as TodoActionTypes from '../actiontypes/todo';

const initialState = {
  todos: [
    {
      id: '10000',
      description: 'hello',
      completed: true
    },
    {
      id: '10001',
      description: 'jack',
      completed: false
    },
    {
      id: '10002',
      description: 'science hw',
      completed: true
    },
    {
      id: '10003',
      description: 'math test',
      completed: true
    }
  ]
};

let apiId = 0;

const findItemIndexById = (id, list) => list.findIndex(item => item.id === id);

// TODO: add axios and post/get
// Should be a pure function (does not mutate the state)
export default function Todo(state = initialState, action) {
  // We use action.type seen in the type in actions/player.js and name aswell
  switch (action.type) {
    case TodoActionTypes.ADD_TODO:
      const addPlayerList = [
        ...state.todos,
        {
          id: apiId.toString(),
          description: action.description,
          completed: action.completed
        }
      ];
      apiId += 1;
      return {
        ...state,
        todos: addPlayerList
      };

    case TodoActionTypes.REMOVE_TODO:
      const todoIndex = findItemIndexById(action.id, state.todos);
      const removePlayerList = [
        ...state.todos.slice(0, todoIndex),
        ...state.todos.slice(todoIndex + 1)
      ];
      return {
        ...state,
        todos: removePlayerList
      };

    case TodoActionTypes.UPDATE_TODO_DESCRIPTION:
      const updatePlayerDescriptionList = state.todos.map((todo, index) => {
        if (index === findItemIndexById(action.id, state.todos)) {
          return {
            ...todo,
            description: action.description
          }
        }
        return todo;
      });
      return {
        ...state,
        todos: updatePlayerDescriptionList
      };

    case TodoActionTypes.UPDATE_TODO_COMPLETED:
      const updatePlayerCompletedList = state.todos.map((todo, index) => {
        if (index === findItemIndexById(action.id, state.todos)) {
          return {
            ...todo,
            completed: action.completed,
          }
        }
        return todo;
      });
      return {
        ...state,
        todos: updatePlayerCompletedList
      };

    default:
      return state;
  }
}
