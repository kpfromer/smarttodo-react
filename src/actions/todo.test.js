import * as TodoActionCreators from './todo';
import { CALL_API } from "../middleware/api/api";
import uuid from 'uuid/v4';

jest.mock('uuid/v4', () => jest.fn());

describe('todo actions', () => {
  let action;

  afterEach(() => {
    uuid.mockClear();
  });

  describe('fetchTodos', () => {
    beforeEach(() => {
      action = TodoActionCreators.fetchTodos();
    });

    it('creates an action to GET todos', () => {
      expect(action[CALL_API]).toMatchSnapshot();
    });
    it('maps todos\' _id from api response to id', () => {
      const todosFromApi = [
        {
          _id: 'a mongo id',
          description: 'hello'
        },
        {
          _id: 'gibberish',
          description: 'world'
        }
      ];
      expect(action[CALL_API].mapResponse(todosFromApi)).toMatchSnapshot();
    });
  });

  describe('addTodo', () => {
    let description, completed, mockUuid;

    beforeEach(() => {
      description = 'hello world';
      completed = true;
      mockUuid = 'mock uuid';

      uuid.mockReturnValue(mockUuid);

      action = TodoActionCreators.addTodo(description, completed);
    });

    it('sends a tempId for the new todo', () => {
      expect(action.tempId).toBe(mockUuid);
      expect(uuid).toHaveBeenCalled();
    });
    it('sends the todo', () => {
      expect(action.todo).toMatchSnapshot();
    });

    it('creates an api action to POST a todo', () => {
      expect(action[CALL_API]).toMatchSnapshot();
    });
  });

  describe('updateTodo', () => {
    let id, prevTodo, newTodo;
    beforeEach(() => {
      id = '1';
      prevTodo = {
        id,
        description: 'old description',
        completed: false
      };
      newTodo = {
        id,
        description: 'NEW! description',
        completed: true
      };
      action = TodoActionCreators.updateTodo(prevTodo, newTodo, id);
    });

    it('updates old todo by id using new todo', () => {
      expect(action.id).toBe(id);
      expect(action.todo).toMatchSnapshot();
    });
    it('can revert back to old todo', () => {
      expect(action.revert).toMatchSnapshot();
    });
    it('creates an action to PUT a new todo', () => {
      expect(action[CALL_API]).toMatchSnapshot();
    });
  });

  describe('removeTodo', () => {
    let todo, position, id;
    beforeEach(() => {
      todo = {
        description: 'to be removed!',
        completed: true
      };
      position = 20;
      id = '4124id';
      action = TodoActionCreators.removeTodo(todo, position, id);
    });

    it('removes todo by id', () => {
      expect(action.id).toBe(id);
    });
    it('can revert back to old todo at the same position', () => {
      expect(action.revert).toMatchSnapshot();
    });
    it('creates an action to DELETE a todo', () => {
      expect(action[CALL_API]).toMatchSnapshot();
    });
  });
});