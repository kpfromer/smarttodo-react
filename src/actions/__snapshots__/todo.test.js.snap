// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todo actions addTodo creates an api action to POST a todo 1`] = `
Object {
  "data": Object {
    "todos": Array [
      Object {
        "completed": true,
        "description": "hello world",
      },
    ],
  },
  "endpoint": "/todo",
  "method": "POST",
  "types": Array [
    "todo/ADD_TODO",
    "todo/ADD_TODO_SUCCESS",
    "todo/ADD_TODO_FAILURE",
  ],
}
`;

exports[`todo actions addTodo creates an api action to add a todo 1`] = `
Object {
  "data": Object {
    "todos": Array [
      Object {
        "completed": true,
        "description": "hello world",
      },
    ],
  },
  "endpoint": "/todo",
  "method": "POST",
  "types": Array [
    "todo/ADD_TODO",
    "todo/ADD_TODO_SUCCESS",
    "todo/ADD_TODO_FAILURE",
  ],
}
`;

exports[`todo actions addTodo sends the todo 1`] = `
Object {
  "completed": true,
  "description": "hello world",
}
`;

exports[`todo actions fetchTodos creates an action to GET todos 1`] = `
Object {
  "endpoint": "/todo",
  "mapResponse": [Function],
  "method": "GET",
  "types": Array [
    "todo/REQUEST_TODOS",
    "todo/RECIEVE_TODOS",
    "todo/REQUEST_FAILURE_TODOS",
  ],
}
`;

exports[`todo actions fetchTodos maps todos' _id from api response to id 1`] = `
Array [
  Object {
    "description": "hello",
    "id": "a mongo id",
  },
  Object {
    "description": "world",
    "id": "gibberish",
  },
]
`;

exports[`todo actions removeTodo can revert back to old todo at the same position 1`] = `
Object {
  "position": 20,
  "todo": Object {
    "completed": true,
    "description": "to be removed!",
  },
}
`;

exports[`todo actions removeTodo creates an action to DELETE a todo 1`] = `
Object {
  "endpoint": "/todo/4124id",
  "method": "DELETE",
  "types": Array [
    "todo/REMOVE_TODO",
    "todo/REMOVE_TODO_SUCCESS",
    "todo/REMOVE_TODO_FAILURE",
  ],
}
`;

exports[`todo actions updateTodo can revert back to old todo 1`] = `
Object {
  "todo": Object {
    "completed": true,
    "description": "NEW! description",
  },
}
`;

exports[`todo actions updateTodo creates an action to PUT a new todo 1`] = `
Object {
  "data": Object {
    "completed": true,
    "description": "NEW! description",
  },
  "endpoint": "/todo/1",
  "method": "PUT",
  "types": Array [
    "todo/UPDATE_TODO",
    "todo/UPDATE_TODO_SUCCESS",
    "todo/UPDATE_TODO_FAILURE",
  ],
}
`;

exports[`todo actions updateTodo updates old todo by id using new todo 1`] = `
Object {
  "completed": true,
  "description": "NEW! description",
  "id": "1",
}
`;
